### PR TITLE
Fix typescript definitions

### PR DESCRIPTION
**Pull request by astegmaier**
### Context
As discussed in #4355 #4326 and #4453, the current `handsontable.d.ts` file included in this project has a number of problems:
- It's impossible (as far as I can tell) to get at a number of types including `gridSettings`, `plugins` from outside of the core Handsontable class. For example, you can't annotate the object you are passing into the `Handsontable` constructor to define grid settings with the `gridSettings` type (to get the benefits of strong type checking).
- You can't extend the classes (such as`Editors`) that are defined in `handsontable.d.ts`
- As @Narazaka noted in #4355, the export style of Handsontable.d.ts has been wrong since [9535a33](https://github.com/handsontable/handsontable/commit/9535a33c1998ec9f34cd47397e3d52257024ebd3), which broke a common way of importing Handsontable with typescript (`import Handsontable = require("handsontable”);`)

This change modifies the way that `Handsontable.d.ts` is structured so that the `Handsontable` export is a both a class that can be instantiated (e.g. `new Handsontable(…)`) as well as a namespace that gives access to all the related helper classes and interfaces (e.g. `Handonstable.GridSettings`, etc.). It is modeled after the structure of the at [@types/handsontable](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/handsontable/index.d.ts) definitions. It also adds a few missing type definitions including `TEXTAREA`, `TEXTAREA_PARENT` and `textareaStyle` to the `Text` editor class, and `Hooks` to the  `GridSettings` interface.

### How has this been tested?
I’ve used the modified type definitions in the context of a typescript project that exercises much of the functionality of Handsontable, including defining custom editors, renderers, and passing settings objects into the Handsontable constructor. I’ve also run the test suite and added tests to exercise this functionality.

### Types of changes

- Bug fix (non-breaking change which fixes an issue)
- New feature or improvement (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

_Note_: I'm not 100% familiar with all the different ways that it is possible to import Handsontable through typescript, but this change restores things to the way they were before [9535a33](https://github.com/handsontable/handsontable/commit/9535a33c1998ec9f34cd47397e3d52257024ebd3), so the following styles should work fine:

`import * as Handsontable from 'handsontable'` and
`import Handsontable = require("handsontable”);`

The following style, which seemed to work in some cases after [9535a33](https://github.com/handsontable/handsontable/commit/9535a33c1998ec9f34cd47397e3d52257024ebd3), will not always work:

`import Handsontable from 'handsontable'`

### Related issue(s):
1. #4355
2. #4326
3. #4453

### Checklist:

- [x] My code follows the code style of this project,
- [x] My change requires a change to the documentation.
None of the documentation currently mentions typescript so technically no changes are _required_, but it would probably be a good idea to describe the correct way to import handsontable into a typescript project and extend `editors` properly with typescript.

### CODE REVIEW

**Comment by AMBudnik** [comment]
I can see that all Contributing guidelines have been achieved. I am marking the PR as `Ready for review`.

Thank you

File: test/types/handsontable-tests.ts [hunk 1]
```diff
@@ -490,3 +491,21 @@ function test_HandsontableMethods() {
...
+    this.TEXTAREA.className = 'handsontableInput';
+    this.textareaStyle =  this.TEXTAREA.style;
+    this.textareaStyle.width = '0';
+    this.textareaStyle.height = '0';
+
+    //replace textarea with password input
+    Handsontable.dom.empty(this.TEXTAREA_PARENT);
+    this.TEXTAREA_PARENT.appendChild(this.TEXTAREA);
+  }
+}
```
Comment (panesofglass) [inline test/types/handsontable-tests.ts]: This is _so_ nice! Thank you!

**Comment by renehamburger** [comment]
Thanks so much, @astegmaier!

@AMBudnik, please consider merging this soon. This is a crucial fix for being able to use Handsontable's own type definitions in any TypeScript project.

**Comment by dmattsson** [comment]
Hi. Is there any ETA on when this could be merged? I'm just trying out handson-table for an angular4 app but got stuck due to this issue. Trying to figure out if I need to spend time on a workaround or if I should wait for the merge.

Many thanks!

btw - came from here:
https://github.com/valor-software/ng2-handsontable/issues/479

**Comment by AMBudnik** [comment]
Hi @dmattsson 

usually, we test PRs on Fridays. 
I will ask our devs to take a look at this PR. @astegmaier has never shared a code that wasn't worth trying but we just have a busy schedule so checking more than 100 lines of code can be rescheduled from one Friday to another.

**Comment by EricTran5791** [comment]
Hoping this gets merged soon. We're considering the pro version and this is a deciding factor.

**Comment by budnix** [comment]
Thanks, @astegmaier for PR! It looks good to me. 

Additionally, I've found that Handsontable is exported differently (depends on for what environment it has been build for) which causes this misunderstanding. I've opened related issue https://github.com/handsontable/handsontable/issues/4605.

**Comment by renehamburger** [comment]
@budnix, when are you planning your next release? It would be great to have this fix available soon.

**Comment by AMBudnik** [comment]
Hi @astegmaier 

the optimistic date is 29th of November, but we are still waiting for one fix so it can take few more days.

**Comment by clairepham91** [comment]
Could you please add licenseKey to handsontable.d.ts for pro version users?

**Pull request merged**